source_control/git.py

```diff
@@ -740,91 +756,107 @@ def main():
-        module.exit_json(changed=False, before=before, after=before)
-    else:
-        # else do a pull
-        local_mods = has_local_mods(module, git_path, dest, bare)
-        before = get_version(module, git_path, dest)
-        if local_mods:
-            # failure should happen regardless of check mode
-            if not force:
-                module.fail_json(msg="Local modifications exist in repository (force=no).")
-            # if force and in non-check mode, do a reset
-            if not module.check_mode:
-                reset(git_path, module, dest)
-        # exit if already at desired sha version
-        set_remote_url(git_path, module, repo, dest, remote)
-        remote_head = get_remote_head(git_path, module, dest, version, remote, bare)
-        if before == remote_head:
-            if local_mods:
-                module.exit_json(changed=True, before=before, after=remote_head,
-                    msg="Local modifications exist")
-            elif is_remote_tag(git_path, module, dest, repo, version):
-                # if the remote is a tag and we have the tag locally, exit early
-                if version in get_tags(git_path, module, dest):
-                    repo_updated = False
-            else:
-                # if the remote is a branch and we have the branch locally, exit early
-                if version in get_branches(git_path, module, dest):
-                    repo_updated = False
-        if repo_updated is None:
-            if module.check_mode:
-                module.exit_json(changed=True, before=before, after=remote_head)
-            fetch(git_path, module, repo, dest, version, remote, bare, refspec)
-            repo_updated = True
-
-    # switch to version specified regardless of whether
-    # we got new revisions from the repository
-    if not bare:
-        switch_version(git_path, module, dest, remote, version, verify_commit)
 
-    # Deal with submodules
-    submodules_updated = False
-    if recursive and not bare:
-        submodules_updated = submodules_fetch(git_path, module, remote, track_submodules, dest)
-
-        if module.check_mode:
-            if submodules_updated:
-                module.exit_json(changed=True, before=before, after=remote_head, submodules_changed=True)
-            else:
-                module.exit_json(changed=False, before=before, after=remote_head)
-
-        if submodules_updated:
-            # Switch to version specified
-            submodule_update(git_path, module, dest, track_submodules)
+    if umask and not isinstance(umask, int):
+        try:
+            umask = int(umask, 8)
+        except Exception:
+            module.fail_json(msg="umask must be an octal integer",
+                    details=str(sys.exc_info()[1]))
 
-    # determine if we changed anything
-    after = get_version(module, git_path, dest)
+    old_umask = None
+    if umask != None:
+      old_umask = os.umask(umask)
```

Comment:
I lifted this code and the umask handling code above from `pip.py`. It might be nice to generalize it so that it doesn't have to be repeated across modules dealing with `umask`.